Flesh out documentation (#12)
lucasmlofaro/cabinets: README.md

```python
import cabinets

new_obj = cabinets.read('test.json')
```

That's it! The file is *loaded* and *parsed* in just one line.

### Write a file

`cabinets` also supports creating files. We can rewrite the first example using
only `cabinets`.

```python
import cabinets

obj = {'test': 1}
cabinets.create('test.json', obj)

new_obj = cabinets.read('test.json')

assert new_obj == obj
```

### Reading and Writing with Other Protocols

> In the previous examples we saw `read` and `write` operations on local files using the
> default *file* protocol. However, one of the main benefits of using `cabinets` is the
> flexibility to read and write files using multiple different protocols, with a common
> interface.

Reviewer: ```suggestion
Using `cabinets` allows you to interact with multiple file storage protocols
depending on the URI you specify. In the previous examples, we used
`read()` and `write()` to operate within our local file system; that's
because `cabinets` assumes we're using the `file://` protocol by default.
Luckily, accessing other storage systems is just as easy!
```

> In addition to supporting the local filesystem protocol, `cabinets`
> natively supports file operations on AWS S3:

Reviewer: ```suggestion
For example, operating on a file on AWS S3 is done exactly the same way:
```

- Python Pickle (`.pickle`)
- CSV *(beta)* (`.csv`)

```python
import cabinets

# .foo file in local filesystem
local_foo_data = cabinets.read('file://test.foo')

# .foo file in S3
s3_foo_data = cabinets.read('s3://test.foo')
```

## Protocol Configuration

> Some storage platform protocols may require some configuration parameters to be set
> before they can be used. Each `Cabinet` subclass can expose
> a `set_configuration(**config)` classmethod to take care of any required initial setup.

Reviewer: ```suggestion
Some storage platform protocols may require additional configuration parameters to be set
before they can be used. Each `Cabinet` subclass can expose a `set_configuration(**config)`
class method to take care of any required initial setup.
```

```

The first call that occurs will be `FooCabinet._read_content('foo.json)`, and that
result is then parsed by the `JSONParser` before being returned.

> **NOTE**: In order for the protocols to be registered, the class definition must be
> run at least once. Make sure the modules where your custom `Cabinet` classes are defined
> are imported somewhere before they are used, OR use the built in [Plugin](#plugins) system.

### Adding Parsers

`cabinets` also supports custom extension parsing in the exact same way:

```python
from cabinets.parser import Parser, register_extensions

@register_extensions('bar')
class BarParser(Parser):
    @classmethod
    def load_content(cls, content: bytes):
        # Parse bytes from "bar" file format into a Python object
        ...

    @classmethod
    def dump_content(cls, data: Any):
        # Dump a Python object into bytes in the "bar" file format
        ...
```

Now if we redo our above example using the `.bar` extension:

```python
from ... import FooCabinet, BarParser  # ensure FooCabinet and BarParser are loaded

cabinets.read('foo://example.bar')
```

This statement is roughly equivalent to:

```python

> BarParser.load_content(FooCabinet._read_content('foo.bar'))

Reviewer: I don't think `_read_content()` is supposed to protected here
```suggestion
BarParser.load_content(FooCabinet.read_content('foo.bar'))
```

Author: Right now the `Cabinet` class methods are protected. They probably shouldn't be.

Author: Now this is right.

```

and should return a Python object from your `Foo` cabinet, using your `Bar` parser!

### Loading Plugins

As mentioned in the example above, your custom `Cabinet` and `Parser` classes must be

> executed in order to be added to the internal cache `cabinets` uses for extension and
> protocol lookup. If your custom classes are imported before any `cabinets` functions are

Reviewer: Keep the respective order consistent.
```suggestion
executed in order to be added to the internal cache `cabinets` uses for protocol and
extension lookup. If your custom classes are imported before any `cabinets` functions are
```

> used then this won't be an issue. However, in many use cases there is no reason to

Reviewer: ```suggestion
use then, this won't be an issue. However, in many use cases there is no reason to
```

import those classes aside from usage with `cabinets` functions. Instead of requiring
each class to be imported manually at the start of your program,
`cabinets` can search a specified path for new `Cabinet` and `Parser` classes, and load
them automatically.

> If the environment variable `PLUGIN_PATH` is specified, `cabinets`
> will search that path for subdirectories called `cabinet` and `parser`. Modules residing
> within those directories will be searched for `Cabinet` and `Parser` subclasses,
> respectively.

Reviewer: ```suggestion
Specifying the `PLUGIN_PATH` environment variable will cause `cabinets` to search
for subdirectories called `cabinet` and `parser` in that path. Modules residing
within those directories will be searched for `Cabinet` and `Parser` subclasses
respectively.
```

```
└─ PLUGIN_PATH
    |
    └───cabinet

>     │   │   custom_cabinet.py   
>     └───parser
>     │   │   custom_parser_1.py
>     │   │   custom_parser_2.py

Reviewer: ```suggestion
    │   │   foo_cabinet.py
    └───parser
    │   │   bar_parser.py
    │   │   baz_parser.py
```

```

> If the above `FooCabinet` and `BarParser` classes are placed in
> `custom_cabinet.py` and `custom_parser_1.py` (or `custom_parser_2.py`) they will be
> loaded and registered to their specified protocols/extensions without needing to be
> referenced anywhere else in the program.

Reviewer: ```suggestion
If the above `FooCabinet` and `BarParser` classes are placed in `foo_cabinet.py`
and `bar_parser.py`, they will be loaded and registered to their specified
cache without needing to be referenced anywhere else in the program.
```

## Contributing

This package is open source (see [LICENSE](LICENSE)), so please feel free
to [contribute](https://opensource.guide/how-to-contribute/)

> by submitting a pull request, creating an issue, or contacting the authors directly.

Reviewer: Are there settings we need to enable to allow people to push branches and submit PRs? Remember you couldn't push until I made you a contributor.

### Authors and Contributors:

> * Lucas Lofaro *(Co-Author)*: [lucasmlofaro@gmail.com](mailto:lucasmlofaro@gmail.com)
> * Sam Hollenbach *(
>   Co-Author)*: [samhollenbach@gmail.com](mailto:samhollenbach@gmail.com) 

Reviewer: ```suggestion
- Lucas Lofaro *(Co-Author)*: [lucasmlofaro@gmail.com](mailto:lucasmlofaro@gmail.com)
- Sam Hollenbach *(Co-Author)*: [samhollenbach@gmail.com](mailto:samhollenbach@gmail.com) 
```
